source/cbor.c

```diff
@@ -0,0 +1,774 @@
+    }                                                                                                                  \
+    struct aws_cbor_encoder *encoder = s_cbor_encoder_from_capsule(py_capsule);                                        \
+    if (!encoder) {                                                                                                    \
+        return NULL;                                                                                                   \
+    }
+
+PyObject *aws_py_cbor_encoder_get_encoded_data(PyObject *self, PyObject *args) {
+    (void)self;
+    PyObject *py_capsule;
+    if (!PyArg_ParseTuple(args, "O", &py_capsule)) {
+        return NULL;
+    }
+    struct aws_cbor_encoder *encoder = s_cbor_encoder_from_capsule(py_capsule);
+    if (!encoder) {
+        return NULL;
+    }
+    struct aws_byte_cursor encoded_data = aws_cbor_encoder_get_encoded_data(encoder);
+    if (encoded_data.len == 0) {
+        /* TODO: probably better to be empty instead of None?? */
+        Py_RETURN_NONE;
+    }
+    return PyBytes_FromStringAndSize((const char *)encoded_data.ptr, encoded_data.len);
+}
+
+#define S_ENCODER_WRITE_PYOBJECT(ctype, py_conversion, field)                                                          \
+    static PyObject *s_cbor_encoder_write_pyobject_as_##field(struct aws_cbor_encoder *encoder, PyObject *py_object) { \
+        ctype data = py_conversion(py_object);                                                                         \
+        if (PyErr_Occurred()) {                                                                                        \
+            return NULL;                                                                                               \
+        }                                                                                                              \
+        aws_cbor_encoder_write_##field(encoder, data);                                                                 \
+        Py_RETURN_NONE;                                                                                                \
+    }
+
+S_ENCODER_WRITE_PYOBJECT(uint64_t, PyLong_AsUnsignedLongLong, uint)
+S_ENCODER_WRITE_PYOBJECT(uint64_t, PyLong_AsUnsignedLongLong, negint)
+S_ENCODER_WRITE_PYOBJECT(double, PyFloat_AsDouble, float)
+S_ENCODER_WRITE_PYOBJECT(struct aws_byte_cursor, aws_byte_cursor_from_pybytes, bytes)
+S_ENCODER_WRITE_PYOBJECT(struct aws_byte_cursor, aws_byte_cursor_from_pyunicode, text)
+S_ENCODER_WRITE_PYOBJECT(bool, PyObject_IsTrue, bool)
+
+S_ENCODER_WRITE_PYOBJECT(uint64_t, PyLong_AsUnsignedLongLong, array_start)
+S_ENCODER_WRITE_PYOBJECT(uint64_t, PyLong_AsUnsignedLongLong, map_start)
+S_ENCODER_WRITE_PYOBJECT(uint64_t, PyLong_AsUnsignedLongLong, tag)
+
+PyObject *aws_py_cbor_encoder_write_unsigned_int(PyObject *self, PyObject *args) {
+    PyObject *pylong;
+    S_ENCODER_METHOD_START("O", &pylong);
```

Comment:
the only thing that changes is the last part of s_cbor_encoder_write_pyobject_as? maybe more more of that into macro and let macro just call correct s_cbor_encoder_write_pyobject_as?

having a macro thats just a partial method kinda scares me